File: app/Http/Controllers/HomeController.php
<?php

namespace App\Http\Controllers;

use App\Activite;
use App\Actualite;
use Illuminate\Http\Request;
use App\galerie;
use App\LieuCours;
use App\NiveauScolaire;
use App\Prof;
use App\TypeActivite;
use App\TypeCours;
use App\User;
use App\Video;
use App\Ville;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Routing\Controller;
use Tymon\JWTAuth\Facades\JWTAuth;

class HomeController extends Controller
{
    /**
     * @var User
     */
    protected  $users;
    protected  $profs;
    protected  $typeActivites;
    protected  $activites;
    protected  $typeCours;
    protected  $lieuCours;
    protected  $niveauScolaires;
    protected  $videos;
    protected  $villes;

    public function __construct(
        User $users,
        Prof $profs,
        TypeActivite $typeActivites,
        Activite $activites,
        TypeCours $typeCours,
        LieuCours $lieuCours,
        NiveauScolaire $niveauScolaires,
        Video $videos,
        Ville $villes
        )
    {
        $this->users = $users;
        $this->profs = $profs;
        $this->typeActivites = $typeActivites;
        $this->activites = $activites;
        $this->typeCours = $typeCours;
        $this->lieuCours = $lieuCours;
        $this->niveauScolaires = $niveauScolaires;
        $this->videos = $videos;
        $this->villes = $villes;
    }


    public function index()
    {
        // $activites = $this->activites
        //     ->get()
        //     ;

        $users = $this->users
            ->get()
            ;

        $videos = $this->videos
            ->orderBy('order', 'asc')
            ->take(3)
            ->get()
            ;

        $activites = $this->activites
            ->where('idTypeActivite', 1)
            ->with(['typeActivite'])
            ->get();

        $villes = $this->villes
            ->get();

        // $typeActivites = $this->typeActivites
        //     ->where('active', true)
        //     ->with(['activites'])
        //     ->get()
        //     ;

        $user = $user = JWTAuth::user();

        $profs = $this->profs
            ->with(['user'])
            ->orderBy('note', 'desc')
            ->take(3)
            ->get()
            ;

        return view('page/home/home', compact('users', 'profs', 'videos', 'activites', 'villes', 'user'));
    }

    public function profs(Request $request
    , int $startIndex, int $pageSize, int $typeActivite, int $activite, int $typeCours, int $lieuCours, int $niveauScolaire, string $prof, int $ville)
    {
        // $activite = $request->input('activite');
        // $typeCours = $request->input('typeCours');
        // $niveauScolaire = $request->input('niveauScolaire');
        //->where(1, '<>', 1);
        $activites = $this->activites->get();
        $typeCourss = $this->typeCours->get();
        $niveauScolaires = $this->niveauScolaires->get();

        $matchThese = [ ];

        // if ($prof != '*') {
        //     array_push($matchThese, ['user.nom', 'LIKE', "%{$prof}%"]);
        // }

        if ($typeActivite != 0) {
            array_push($matchThese, ['idsTypeActivites', 'LIKE', "%;{$typeActivite};%"]);
        }

        if ($activite !== 0) {
            array_push($matchThese, ['idsActivites', 'LIKE', "%;{$activite};%"]);
        }

        if ($typeCours != 0) {
            array_push($matchThese, ['idsTypeCours', 'LIKE', "%;{$typeCours};%"]);
        }

        // if ($lieuCours != 0) {
        //     // $q->orWhere('idsLieuCours', 'LIKE', "%;{$lieuCours};%");
        //     array_push($matchThese, ['idsLieuCours', 'LIKE', "%;{$lieuCours};%"]);
        // }

        if ($niveauScolaire != 0) {
            // $q->orWhere('idsNiveauScolaires', 'LIKE', "%;{$niveauScolaire};%");
            array_push($matchThese, ['idsNiveauScolaires', 'LIKE', "%;{$niveauScolaire};%"]);
        }

        $q = $this->profs
            ->where($matchThese)
            ->whereHas('user', function (Builder $query) use ($prof, $ville) {
                $query->where([
                    $prof != '*' ? ['nom', 'LIKE', "%{$prof}%"] : ['1','1'],
                    $ville != 0 ? ['idVille', $ville] : ['1','1']
                ]);
            })
            ->with(['user'])
            ->orderBy('note', 'desc');

        $count = $q->count();

        $profs = $q
            // ->skip($startIndex)
            // ->take($pageSize)
            ->paginate($pageSize)
            // ->appends(request()->except('page'))
            // ->get()
            ;

        return view('page/profs', compact('profs', 'activites', 'count', 'typeCourss', 'niveauScolaires'));
    }

    // public function profs0(Request $req)
    // {
    //     // $activite = $request->input('activite');
    //     // $typeCours = $request->input('typeCours');
    //     // $niveauScolaire = $request->input('niveauScolaire');
    //     //->where(1, '<>', 1);
    //     $activites = $this->activites->get();
    //     $typeCourss = $this->typeCours->get();
    //     $niveauScolaires = $this->niveauScolaires->get();

    //     $matchThese = [ ];

    //     if ($prof != '') {
    //         array_push($matchThese, ['user.nom', 'LIKE', "%{$prof}%"]);
    //     }

    //     if ($typeActivite != 0) {
    //         array_push($matchThese, ['idsTypeActivites', 'LIKE', "%;{$typeActivite};%"]);
    //     }

    //     if ($activite !== 0) {
    //         array_push($matchThese, ['idsActivites', 'LIKE', "%;{$activite};%"]);
    //     }

    //     if ($typeCours != 0) {
    //         array_push($matchThese, ['idsTypeCours', 'LIKE', "%;{$typeCours};%"]);
    //     }

    //     // if ($lieuCours != 0) {
    //     //     // $q->orWhere('idsLieuCours', 'LIKE', "%;{$lieuCours};%");
    //     //     array_push($matchThese, ['idsLieuCours', 'LIKE', "%;{$lieuCours};%"]);
    //     // }

    //     if ($niveauScolaire != 0) {
    //         // $q->orWhere('idsNiveauScolaires', 'LIKE', "%;{$niveauScolaire};%");
    //         array_push($matchThese, ['idsNiveauScolaires', 'LIKE', "%;{$niveauScolaire};%"]);
    //     }

    //     $q = $this->profs
    //         ->where($matchThese)
    //         ->with(['user'])
    //         ->orderBy('note', 'desc');

    //     $count = $q->count();

    //     $profs = $q
    //         // ->skip($startIndex)
    //         // ->take($pageSize)
    //         ->paginate($pageSize)
    //         // ->appends(request()->except('page'))
    //         // ->get()
    //         ;

    //     return view('page/profs', compact('profs', 'activites', 'count', 'typeCourss', 'niveauScolaires'));
    // }

    public function prof(int $id)
    {

        $model = $this->profs->with(['user'])->find($id);

        $typeActivites = $this->typeActivites->get();
        $activites = $this->activites->get();

        $typeCours = $this->typeCours->get();
        $lieuCours = $this->lieuCours->get();
        $niveauScolaires = $this->niveauScolaires->get();

        return view('page/prof', compact('model', 'typeActivites', 'activites', 'typeCours', 'lieuCours', 'niveauScolaires'));
    }

    public function videos(int $startIndex, int $pageSize)
    {

        $videos = $this->videos
            ->orderBy('order', 'asc')
            // ->skip($startIndex)
            // ->take($pageSize)
            ->paginate($pageSize)
            ->appends(request()->except('page'))
            // ->get()
            ;

        return view('page/videos', compact('videos'));
    }

    public function video(int $id)
    {
        $model = $this->videos
            ->where('id', $id)
            ->get()
            ;

        return view('page/video', compact('model'));
    }
}

File: Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Models;
using Api.Providers;
using Microsoft.AspNetCore.WebUtilities;
using System.Text;

namespace Controllers
{
    [Route("api/[controller]/[action]")]
    [ApiController]
    public class HomeController : ControllerBase
    {
        protected readonly MyContext _context;

        public HomeController(MyContext context)
        {
            _context = context;
        }

        [HttpGet]
        public async Task<IActionResult> Welcome()
        {
            var videos = await _context.Videos.Take(3).OrderBy(e => e.Order).ToListAsync();

            var activites = await _context.Activites.Where(e => e.IdTypeActivite == 1)
            .Select(e => new
            {
                Id = e.Id,
                Nom = e.Nom,
                NomAr = e.NomAr,
                imageUrl = e.ImageUrl,
                typeActivite = new
                {
                    nom = e.TypeActivite.Nom,
                    nomAr = e.TypeActivite.NomAr,
                },
            })
            .ToListAsync()
            ;

            var villes = await _context.Villes.ToListAsync();

            var profs = await _context.Profs.OrderByDescending(e => e.Note)
            .Take(5)
            .Select(e => new
            {
                note = e.Note,
                intro = e.Intro,
                user = new
                {
                    id = e.User.Id,
                    imageUrl = e.User.ImageUrl,
                    Nom = e.User.Nom,
                    Prenom = e.User.Prenom,
                }
            })
            .ToListAsync();

            return Ok(new { videos, activites, villes, profs });
        }

        [HttpGet]
        public async Task<IActionResult> SearchProfs(
          [FromQuery] int startIndex = 0,
          [FromQuery] int pageSize = 10,
          [FromQuery] int typeActivite = 0,
          [FromQuery] int activite = 0,
          [FromQuery] int typeCours = 0,
          [FromQuery] int lieuCours = 0,
          [FromQuery] int niveauScolaire = 0,
          [FromQuery] string prof = "",
          [FromQuery] int ville = 0
        )
        {
            var g = prof == null;

            var q = _context.Profs
              .Where(e => typeActivite == 0 ? true : e.IdsTypeActivites.Contains($";{typeActivite};"))
              .Where(e => activite == 0 ? true : e.IdsActivites.Contains($";{activite};"))
              .Where(e => typeCours == 0 ? true : e.IdsTypeCours.Contains($";{typeCours};"))
              .Where(e => niveauScolaire == 0 ? true : e.IdsNiveauScolaires.Contains($";{niveauScolaire};"))
              .Where(e => prof == null || prof == "" ? true : e.User.Nom.ToLower().Contains(prof.ToLower()) || e.User.Prenom.ToLower().Contains(prof.ToLower()))
              .Where(e => ville == 0 ? true : e.User.IdVille == ville)
              ;

            var count = await q.CountAsync();

            var list = await q.Skip(startIndex)
              .Take(pageSize)
              .Select(e => new
              {
                  note = e.Note,
                  intro = e.Intro,
                  Description = e.Description,
                  IdsActivites = e.IdsActivites,
                  user = new
                  {
                      id = e.User.Id,
                      imageUrl = e.User.ImageUrl,
                      Nom = e.User.Nom,
                      Prenom = e.User.Prenom,
                  }
              })
              .ToListAsync()
              ;


            return Ok(new { list, count });
        }
    }
}